File: inno/summer-shop-2017/05:07/kthstat/solutions/kthstat_sk_nk.pas
// 20 points (TL) by Sergey Kopeliovich

unit kthstat;

interface

uses graderlib;

procedure Start( nn : longint; var aa : mas );

procedure Add( i, x : longint );
procedure Del( i : longint );

function Get( L, R, x : longint ) : longint;

implementation

var
  i, j, n : longint;
  a : mas;

procedure Start( nn : longint; var aa : mas );
begin
  n := nn;
  for i := 0 to n - 1 do
    a[i] := aa[i];
end;


procedure Add( i, x : longint );
begin
  for j := n downto i + 1 do
    a[j] := a[j - 1];
  a[i] := x;
  inc(n);
end;

procedure Del( i : longint );
begin
  for j := i + 1 to n - 1 do
    a[j - 1] := a[j];
  dec(n);
end;

function Get( L, R, x : longint ) : longint;
var
  cnt : longint;
begin
  cnt := 0;
  for i := L to R do
    if (a[i] <= x) then
      inc(cnt);
  Get := cnt;
end;

begin
end.

File: inno/summer-shop-2017/05:07/kthstat/solutions/kthstat_sk_nk.cpp
// 20 points (TL) by Sergey Kopeliovich

#include "grader.h"

#define maxn (int)1e5

int n, a[maxn];

void Start( int _n, mas _a )
{
  int i;
  n = _n;
  for (i = 0; i < n; i++)
    a[i] = _a[i];
}

void Add( int i, int x )
{
  int j;
  for (j = n; j > i; j--)
    a[j] = a[j - 1];
  a[i] = x, n++;
}

void Del( int i )
{
  int j;
  for (j = i + 1; j < n; j++)
    a[j - 1] = a[j];
  n--;
}

int Get( int L, int R, int x )
{
  int cnt = 0;
  while (L <= R)
    cnt += a[L++] <= x;
  return cnt;
}
